consensus/src/quorum_store/utils.rs

```diff
@@ -467,7 +528,116 @@ impl ProofQueue {
         return_non_full: bool,
         block_timestamp: Duration,
     ) -> (Vec<ProofOfStore>, PayloadTxnsSize, u64, bool) {
-        let mut ret = vec![];
+        let (result, all_txns, unique_txns, is_full) = self.pull_internal(
+            false,
+            excluded_batches,
+            max_txns,
+            max_txns_after_filtering,
+            soft_max_txns_after_filtering,
+            return_non_full,
+            block_timestamp,
+        );
+        let proof_of_stores: Vec<_> = result
+            .into_iter()
+            .map(|item| {
+                let proof = item.proof.clone().expect("proof must exist due to filter");
+                let bucket = proof.gas_bucket_start();
+                counters::pos_to_pull(
+                    bucket,
+                    item.proof_insertion_time
+                        .expect("proof must exist due to filter")
+                        .elapsed()
+                        .as_secs_f64(),
+                );
+                proof
+            })
+            .collect();
+
+        if is_full || return_non_full {
+            counters::BLOCK_SIZE_WHEN_PULL.observe(unique_txns as f64);
+            counters::TOTAL_BLOCK_SIZE_WHEN_PULL.observe(all_txns.count() as f64);
+            counters::KNOWN_DUPLICATE_TXNS_WHEN_PULL
+                .observe((all_txns.count().saturating_sub(unique_txns)) as f64);
+            counters::BLOCK_BYTES_WHEN_PULL.observe(all_txns.size_in_bytes() as f64);
+
+            counters::PROOF_SIZE_WHEN_PULL.observe(proof_of_stores.len() as f64);
+            // Number of proofs remaining in proof queue after the pull
+            self.log_remaining_data_after_pull(excluded_batches, &proof_of_stores);
+        }
+
+        (proof_of_stores, all_txns, unique_txns, !is_full)
+    }
+
+    pub fn pull_batches(
+        &mut self,
+        excluded_batches: &HashSet<BatchInfo>,
+        max_txns: PayloadTxnsSize,
+        max_txns_after_filtering: u64,
+        soft_max_txns_after_filtering: u64,
+        return_non_full: bool,
+        block_timestamp: Duration,
+    ) -> (Vec<BatchInfo>, PayloadTxnsSize, u64) {
+        let (result, all_txns, unique_txns, _) = self.pull_internal(
+            true,
+            excluded_batches,
+            max_txns,
+            max_txns_after_filtering,
+            soft_max_txns_after_filtering,
+            return_non_full,
+            block_timestamp,
+        );
+        let batches = result.into_iter().map(|item| item.info.clone()).collect();
```

Comment:
May be we could add some metrics for the batches pulled here.

Reply:
Yeah, I will add more counters in the next PR. It needs some refactoring too.